pkg/reconciler/configmap/configmap.go

```diff
@@ -0,0 +1,423 @@
+package configmap
+
+import (
+	"context"
+	"fmt"
+	v1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	v13 "k8s.io/api/rbac/v1"
+	"k8s.io/api/rbac/v1alpha1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	v12 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/client-go/tools/record"
+	"regexp"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sort"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+const (
+	//TODO eventually we'll need to decide if we want to make this tuneable
+	contextTimeout           = 300 * time.Second
+	UserConfigMapName        = "jvm-build-config"
+	CacheDeploymentName      = "jvm-build-workspace-artifact-cache"
+	LocalstackDeploymentName = "jvm-build-workspace-localstack"
+	EnableRebuilds           = "enable-rebuilds"
+	ServiceAccountName       = "hacbs-jvm-cache"
+	SystemConfigMapName      = "jvm-build-system-config"
+	SystemConfigMapNamespace = "jvm-build-service"
+	SystemCacheImage         = "image.cache"
+	SystemBuilderImages      = "builder-image.names"
+	SystemBuilderImageFormat = "builder-image.%s.image"
+)
+
+var (
+	RequiredKeys = map[string]bool{SystemCacheImage: true}
+	log          = ctrl.Log.WithName("configmap")
+)
+
+type ReconcileConfigMap struct {
+	client               client.Client
+	scheme               *runtime.Scheme
+	eventRecorder        record.EventRecorder
+	configuredCacheImage string
+	builderImages        *BuilderImageConfig
+}
+
+type BuilderImageConfig struct {
+	Lock   sync.Locker
```

Comment:
@gabemontero can I get your opinion on this bit? Basically the builder images are needed in the dependencybuild reconciler, and it seems really wasteful to have to explcitiy grab the config map there every time we do a build. 

Instead I just have this shared BuilderImageConfig object and access it under a lock.

Reply:
yeah since we are reconciling on the config map it should be in the watch cache and hence grabbing it each time using the controller runtime client is suppose to be a "local fetch / in memory only / cache hit" with no RPC.

general rule of thumb we don't want to synchronize across events like this

Reply:
I have changed this to just get the config map in the db reconciler

Reply:
yep in conjunction with https://github.com/redhat-appstudio/jvm-build-service/pull/105/files#diff-243ebed2765f75e6a54f57167212fefb08c3b2a85967ad2acbc0eb78919019c1R84 looks good  / thanks

marking resolved